### ISSUE TITLE
Make Docker ps more robust to problem containers

### ISSUE THREAD

**Issue by SamSaffron** [moby/moby#28183]
As it stands there are quite a few long standing issues with `docker ps` hanging. 

https://github.com/docker/docker/issues/25993
https://github.com/docker/docker/issues/12606

and so on. 

At the root one of the biggest issues is that if a single container is blocking, you can no longer iterate through the container list: 

can `reducePSContainer` be made more robust so it can skip a container if it is locked for more than N seconds and prints out the name of the problem container. 

This would make it significantly easier to debug this issue and isolate the underlying cause.

**Comment by icecrime** [comment]
Ping @cpuguy83 who I believe was looking into this.

**Comment by cpuguy83** [comment]
Indeed I am. Still some issues to work out but it's coming along.
https://github.com/cpuguy83/docker/compare/container_memdb

**Comment by cpuguy83** [comment]
Basic idea is to strip the container object down to configuration only and move runtime state (e.g. IO streams, log driver, etc) to a separate object.
The container object will be on a transactional in-memory store where when requested the daemon gives copies of the container object rather than a pointer reference, which means there is no longer a need to lock this object. It can be used for reporting at any time without risk of blocking.

State manipulation actions (such as start/stop) will still require a lock, but not on the container object itself. This ensures the daemon can still report status, and potentially even warn the user of a problem.

**Comment by SamSaffron** [comment]
awesome @cpuguy83 !

**Comment by SamSaffron** [comment]
To help debug this, since we are seeing tons of lockups on 1.12.3 I am patching the docker binary with: 

https://gist.github.com/SamSaffron/56a9e47f0c2d4a6e01ee981b30a3b87c

Hopefully this will help me find the culprit backtrace on the problem machines.

**Comment by cpuguy83** [comment]
@SamSaffron If you can send `SIGUSR1` to the daemon and post the backtraces here we can take a look.

**Comment by SamSaffron** [comment]
We had a hang this morning, but I was not around to get the full stack traces... I did however get this via my logging:

```
WARN[63172] Holding mutex for longer than 10 secs: pid 6317
  goroutine 1647633 [running]:
github.com/docker/docker/container.(*State).Lock(0xc820060c80)
    /go/src/github.com/docker/docker/container/state.go:41 +0x8d
github.com/docker/docker/daemon.(*Daemon).StateChanged(0xc82046c000, 0xc8210ac240, 0x40, 0x1e408f0, 0xc, 0x0, 0xc8210ac2c0, 0x40, 0x0, 0x0, ...)
    /go/src/github.com/docker/docker/daemon/monitor.go:62 +0xd86
github.com/docker/docker/libcontainerd.(*container).handleEvent.func1()
    /go/src/github.com/docker/docker/libcontainerd/container_linux.go:184 +0xab
github.com/docker/docker/libcontainerd.(*queue).append.func1(0xc821f41c01, 0xc8212dcb40, 0xc82308a1b0, 0xc8219125a0)
    /go/src/github.com/docker/docker/libcontainerd/queue_linux.go:26 +0x47
created by github.com/docker/docker/libcontainerd.(*queue).append
    /go/src/github.com/docker/docker/libcontainerd/queue_linux.go:28 +0x1da
```

looks like the hang originated from libcontainerd, any tips here?  @cpuguy83

**Comment by cpuguy83** [comment]
@SamSaffron Hard to tell without seeing what's holding the state lock.
Maybe you can have the patch dump the full stack so we can pick through it?

**Comment by SamSaffron** [comment]
I am pretty sure the hang is in:

```
case libcontainerd.StateExitProcess:
		c.Lock()
		defer c.Unlock()
		if execConfig := c.ExecCommands.Get(e.ProcessID); execConfig != nil {
			ec := int(e.ExitCode)
			execConfig.ExitCode = &ec
			execConfig.Running = false
			execConfig.Wait()    // <<< Hangs here
			if err := execConfig.CloseStreams(); err != nil {
				logrus.Errorf("%s: %s", c.ID, err)
			}

			// remove the exec command from the container's store only and not the
			// daemon's store so that the exec command can be inspected.
			c.ExecCommands.Delete(execConfig.ID)
		} else {
			logrus.Warnf("Ignoring StateExitProcess for %v but no exec command found", e)
		}
```

The code is smart, it grabs the backtrace each time you take a mutex and then spits it out when it detects the mutex was not released

Also getting tons of:

```
WARN[13080] containerd: unable to save 212a1d3ca5cf0e751e837a028592d2068499ab02ffcb6b654d726ffa6409f942:fdcc2a7743be8a16a06b3eed5ecf93213157e979730c1d4b1c6693a185bf2324 starttime: open /proc/2020/stat: no such file or directory 
INFO[13080] containerd: 212a1d3ca5cf0e751e837a028592d2068499ab02ffcb6b654d726ffa6409f942:fdcc2a7743be8a16a06b3eed5ecf93213157e979730c1d4b1c6693a185bf2324 (pid 2020) has become an orphan, killing it 
```

This appears fixed in latest libcontainerd afaik. 


My guess is that there is a race conditions where AttachStream never fires copyFunc so `.Done()` is not called.

I wonder if that Wait should have a timeout?

**Comment by cpuguy83** [comment]
Oh you're right, that's from lock that's being held not the waiter.

ping @mlaventure 
Can you have a look at this?
Looks like exec stdio streams having an issue closing in 1.12.3?

**Comment by cpuguy83** [comment]
On master I don't think we are even taking the container lock here, which seems to make sense.

**Comment by mlaventure** [comment]
Yes, the container lock was removed from master recently.

There's still the case of the `execConfig` lock though, it can get stuck until all the exec children have released the IO streams.

**Comment by SamSaffron** [comment]
@mlaventure what if `execConfig.Wait()` had a 10 second timeout, if it fails to get signalled just have it  `CloseStreams` unconditionally and rudely. It seems far less risky.

**Comment by mlaventure** [comment]
@SamSaffron that would break backward compatibility as those processes will probably just die as a result.

Maybe this wait could be moved into a go routine though, would have to look at the implications.

**Comment by SamSaffron** [comment]
I just find the hangs far more severe... I am trying out:

```
                       done := false
			go func() {
				time.Sleep(10 * time.Second)
				if !done {
					logrus.Errorf("%s: Failed to wait for execConfig", c.ID)
					execConfig.Done()
				}
			}()

			execConfig.Wait()

			done = true
```

What is the worst case scenario here? The container is already restarting anyway so they might as well die

**Comment by mlaventure** [comment]
@SamSaffron that case is not only for restarting container.

If the container is restarting because its init process died, this shouldn't block as all processes in that container would have been killed

**Comment by SamSaffron** [comment]
@mlaventure it looks like my patch worked for us on 1.12 

Master and 1.13 no longer lock the container on StateExitProcess however I feel that my fix still should be considered for docker core. 

The container that exhibited the issue has been running for the past 3 days, it experienced the issue 8 hours ago. 

What somehow happened is that we `docker exec`ed something in the container and then dockerd detected that the process exited but AttachStream was somehow stuck waiting on io for exact same process that was execed. 

The patch above protects against this condition.

**Comment by mlaventure** [comment]
@SamSaffron does this process create any other processes that keeps the standard IOs open?

We cannot just cancel the exec after a timeout as that may break other setups that do just what I've described above.

**Comment by SamSaffron** [comment]
The container has only one process (PID 1) this process does not fork or create any other process. 

We have a monitoring piece that runs this every minute (and a few other command) 

```
docker exec container ps aux
```

Trouble is there is a 1.12 bug in libcontainer where it can't properly keep track of short lived execs and causes error log flooding.

I will upgrade this server to 1.13 and add logging around that area to ensure nothing gets stuck in "wait" for longer than N seconds and log if it does. I don't want to be leaking goroutines due to this edge case we have.

**Comment by mlaventure** [comment]
@SamSaffron I was referring to the exec process creating children.

Do you have daemon logs (in debug mode) showing that the exit event for short lived exec are being ignored in libcontainerd as of 1.12.3?

If that's the case, this requires fixing :)

**Comment by cpuguy83** [comment]
I think we should pull in the "don't hold container lock on exec exit" fix for a hypothetical 1.12.x at the very least.

**Comment by itai-codefresh** [comment]
any update about this?
docker 1.12.5 has been released. was this introduced in the newest version?

**Comment by cpuguy83** [comment]
@itai-codefresh there are a couple of fixes in 1.12.4 and 1.12.5.
We don't hold the container lock on exec exit.
Stdio handling is improved (which caused the deadlock while holding the container lock on exec exit)
We now use timeouts on netlink requests and boltdb queries.

So yes, definitely upgrade.

**Comment by fabiokung** [comment]
We are seeing the same lockups on `docker ps` because `(*Daemon).containerStart()` is stuck waiting for containerd to respond and holding the lock.

`(*Daemon).reducePsContainer()` is still trying to grab the container lock on master. @cpuguy83 any progress on your memdb patches?

**Comment by SamSaffron** [comment]
@fabiokung I have not seen these particular lockups in a bit but not entirely sure what we changed to reduce them, short of upgrading. what version are you on?

**Comment by cpuguy83** [comment]
Not much progress of late, sadly.

There are now a couple of places with timeouts applied that help this situation a bit, particularly timeouts on netlink requests.

**Comment by cpuguy83** [comment]
@SamSaffron Glad to hear it's cleared up a bit for you.

**Comment by fabiokung** [comment]
@SamSaffron we are on `v1.12.6`, but the bug causing the lockup during starts in containerd hasn't been released yet (it's on the `1.13.x` tree, but not in `1.13.1-rc1`): docker/containerd#436

On the docker side, it seems valid to hold a lock during `(*Daemon).containerStart()`, so the fix is indeed to not require locking each container on `(*Daemon).reducePsContainer()`.
